components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/DefaultRoleManagementListenerTest.java

```diff
@@ -0,0 +1,224 @@
+import org.testng.annotations.Test;
+import org.wso2.carbon.identity.application.common.IdentityApplicationManagementException;
+import org.wso2.carbon.identity.application.common.model.ServiceProvider;
+import org.wso2.carbon.identity.application.common.model.ServiceProviderProperty;
+import org.wso2.carbon.identity.application.mgt.listener.DefaultRoleManagementListener;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.role.v2.mgt.core.exception.IdentityRoleManagementClientException;
+import org.wso2.carbon.identity.role.v2.mgt.core.exception.IdentityRoleManagementServerException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+public class DefaultRoleManagementListenerTest {
+
+    private DefaultRoleManagementListener defaultRoleManagementListener;
+    private static final String ROLE_NAME = "test_role";
+    private static final String APPLICATION_NAME = "app_name";
+    private static final String APPLICATION_RES_ID = "app_id";
+    private static final String IS_FRAGMENT_APP = "isFragmentApp";
+    private static final String APPLICATION_AUD = "APPLICATION";
+    private static final String ORGANIZATION_AUD = "ORGANIZATION";
+    private static final String TENANT_DOMAIN = "wso2.com";
+
+    @BeforeClass
+    public void setUp() throws Exception {
+
+        defaultRoleManagementListener = spy(new DefaultRoleManagementListener());
+    }
+
+    @DataProvider(name = "fragmentAppPropertyProvider")
+    public Object[][] fragmentAppPropertyProvider() {
+
+        return new Object[][] {
+                {true},
+                {false}
+        };
+    }
+
+    @Test(dataProvider = "fragmentAppPropertyProvider")
+    public void testPreAddRoleForFragmentApp(boolean isFragmentApp) throws Exception {
+
+        try (MockedStatic<ApplicationManagementService> applicationManagementServiceMockedStatic =
+                     mockStatic(ApplicationManagementService.class)) {
+
+            // Creating service provider object.
+            ServiceProvider serviceProvider = new ServiceProvider();
+            serviceProvider.setApplicationName(APPLICATION_NAME);
+            serviceProvider.setApplicationResourceId(APPLICATION_RES_ID);
+
+            // Creating service provider property object.
+            if (isFragmentApp) {
+                ServiceProviderProperty isFragmentAppSpProp = new ServiceProviderProperty();
+                isFragmentAppSpProp.setName(IS_FRAGMENT_APP);
+                isFragmentAppSpProp.setValue(Boolean.TRUE.toString());
+                serviceProvider.setSpProperties(new ServiceProviderProperty[]{isFragmentAppSpProp});
+            } else {
+                serviceProvider.setSpProperties(null);
+            }
```

Comment:
Better to move the app creating also into the data provider. Create two apps 1 -> shraed app other one -> non shared apps, data provider will run in two cases as same isFragmentApp true/false

Then the test method is clean

Reply:
Moved